collector/snapshots.go

```diff
@@ -0,0 +1,302 @@
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type snapshotMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(snapshotStats SnapshotStatDataResponse) float64
+	Labels func(repositoryName string, snapshotStats SnapshotStatDataResponse) []string
+}
+
+type repositoryMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(snapshotsStats SnapshotStatsResponse) float64
+	Labels func(repositoryName string) []string
+}
+
+var (
+	defaultSnapshotLabels      = []string{"repository", "state", "version"}
+	defaultSnapshotLabelValues = func(repositoryName string, snapshotStats SnapshotStatDataResponse) []string {
+		return []string{repositoryName, snapshotStats.State, snapshotStats.Version}
+	}
+	defaultSnapshotShardLabels = append(defaultSnapshotLabels, "metric")
```

Comment:
somehow `metric` is a bit to generic, but I'm struggling to find a better name. This is usually not a problem of the one who's trying to find the label name but a problem of the label itself. The label values simply have too less in common. So we should separate the metrics in
- `snapshot_total_shards`
- `snapshot_failed_shards`
- `snapshot_successful_shards`

Another hint for label design is, that it should make sense to sum over the labels. If it doesn't make sense, it shouldn't be in one metric.

Sorry for coming up with this a bit late.

I'd also suggest to drop the `last_` from all the snapshot metrics. Prometheus always scrapes the latest available data, so this seems a bit redundant